pgwire: trim statements past the conn executor's rewind pos eagerly

asubiotto opened the pull request:
Previously, the conn executor held on to all executed statements in case they
needed to be retried. These would only be discarded once the transaction
committed. This could lead to a memory blowup in some cases.

With the knowledge that statements whose results have been returned cannot be
retried, this commit eagerly discards those statements every time a command is
executed rather than waiting for the transaction to commit.

Release note (bug fix): executing a large number of statements in a transaction
without committing could previously crash a cockroach server, this is now
fixed.

Addresses #47969

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48859)
<!-- Reviewable:end -->

andreimatei reviewed (changes requested):
Please add a test. It's very easy for this new silent logic to be discarded.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-:-M7IjLPC8omHk3--GuZh:bub6hyt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)

___
*[pkg/sql/conn_executor.go, line 1535 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M7Ij5Zl0-hGvMWnaM-u:-M7IgrZ9-1xlfWvshPPg:bd7ltr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9318073c248d3b364ed5244bbd4662e262c29e6/pkg/sql/conn_executor.go#L1535)):*
> ```Go
> 	}
> 
> 	cl := ex.clientComm.LockCommunication()
> ```

Takes some thing thinking to realize that it's OK to put this logic before the update to `txnRewindPos` below. I'd move it after that point.
___
*[pkg/sql/conn_executor.go, line 1536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M7IjdPX4zv3B3X-tIby:-M7IgUH41lqetRR-M5Eh:b9xvd17) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9318073c248d3b364ed5244bbd4662e262c29e6/pkg/sql/conn_executor.go#L1536)):*
> ```Go
> 
> 	cl := ex.clientComm.LockCommunication()
> 	if pos := cl.ClientPos(); pos >= 0 && pos >= ex.extraTxnState.txnRewindPos {
> ```

What's going on here is hard to parse, and I don't understand the `pos >= 0` part. Assigning something to a `canRewind` would probably help. Better yet, I'd reuse `getRewindTxnCapability()` and do the trimming when that returns false. For the case where it returns true, I'd add a `rewindCapability.close()` (besides the `rewindAndUnlock()`).


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Is adding a testing knob to `ExecutorTestingKnobs` to somehow analyze the state of this buffer the right approach to testing this? If not, do you have a suggestion?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-:-M7m_0go23V_EJcRu1Ca:b-u73xt9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)

___
*[pkg/sql/conn_executor.go, line 1536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M7IjdPX4zv3B3X-tIby:-M7mWsTaElKaK25VWNqB:be3rw13) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9318073c248d3b364ed5244bbd4662e262c29e6/pkg/sql/conn_executor.go#L1536)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

What's going on here is hard to parse, and I don't understand the `pos >= 0` part. Assigning something to a `canRewind` would probably help. Better yet, I'd reuse `getRewindTxnCapability()` and do the trimming when that returns false. For the case where it returns true, I'd add a `rewindCapability.close()` (besides the `rewindAndUnlock()`).
</blockquote></details>

SGTM up to `rewindAndUnlock`. Why would we want to do that? This logic should only trim the buffer, not rewind anything.


<!-- Sent from Reviewable.io -->

asubiotto commented:
Added a test and changed the trimming logic to what you suggested, RFAL.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48859)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/conn_executor.go, line 1536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M7IjdPX4zv3B3X-tIby:-M8H1ngUAUKaaSQRUxYN:b-rcl4ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9318073c248d3b364ed5244bbd4662e262c29e6/pkg/sql/conn_executor.go#L1536)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

SGTM up to `rewindAndUnlock`. Why would we want to do that? This logic should only trim the buffer, not rewind anything.
</blockquote></details>

I was just saying to put the `cl.Close()` call in a new `rewindCap.Close()` method.
___
*[pkg/sql/conn_executor_test.go, line 815 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H2Kua3PFTNG-CwXla:-M8H2Kub3OetPEdRHSLj:bke4nbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L815)):*
> ```Go
> }
> 
> // TestTrimFlushedStatements verifies that the conn executor releases statements
> ```

"releases statements" is vague. Say something about that statements buffer.
___
*[pkg/sql/conn_executor_test.go, line 820 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H3WXnEJ5LYtsb9kGB:-M8H3WXnEJ5LYtsb9kGC:bppg7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L820)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	const expectedMaxLen = int32(4)
> ```

explain the `4` pls. Better yet, instead of hardcoding something about the way in which some driver uses pgwire, can you iterate through `buf` inside your filter and assert that you never have more than one `SELECT COUNT(*) FROM test` in there?
___
*[pkg/sql/conn_executor_test.go, line 826 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H4ceDAXmCYv8yXS54:-M8H4ceDAXmCYv8yXS55:b-szf1ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L826)):*
> ```Go
> 		Knobs: base.TestingKnobs{
> 			SQLExecutor: &sql.ExecutorTestingKnobs{
> 				AfterExecuteCmd: func(_ context.Context, buf *sql.StmtBuf) {
> ```

this filter takes everything but the command that was just executed. Pls add it, even if you don't use it yet - particularly since with this patch `buf` will no longer contain the respective command. And, in fact, use the command in this test to short-circuit the filter when the command that was just executed is not your `select count`. Otherwise, you're testing all the random executors used by the server - e.g. the internal ones. If you do that, all the atomic stuff can go away since the filter will be synchronized with the one connection triggering it.
___
*[pkg/sql/conn_executor_test.go, line 844 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H2gDpCWXytJNb2G-r:-M8H2gDpCWXytJNb2G-s:b-vrcujg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L844)):*
> ```Go
> 	require.NoError(t, err)
> 
> 	tx, err := sqlDB.BeginTx(ctx, nil)
> ```

s/BeginTx/Begin and the I won't ask you for a comment on the `nil`
___
*[pkg/sql/conn_executor_test.go, line 848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H33TD8D1mmQOhNdWu:-M8H33TD8D1mmQOhNdWv:b2au3h3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L848)):*
> ```Go
> 
> 	for i := 0; i < 10; i++ {
> 		r, err := tx.Query("SELECT COUNT(*) FROM test")
> ```

do you need the `Query()` instead of `Exec()`?
___
*[pkg/sql/conn_executor_test.go, line 850 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H2lAB3UUryKQsg-sz:-M8H2lAB3UUryKQsg-t-:b-19ih9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L850)):*
> ```Go
> 		r, err := tx.Query("SELECT COUNT(*) FROM test")
> 		require.NoError(t, err)
> 		for r.Next() {
> ```

do you need this `for` loop? I think you can remove it.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
TFTR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-:-M8KCT5u8eDWJAFt8ZYv:bb73zsl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/sql/conn_executor.go, line 1535 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M7Ij5Zl0-hGvMWnaM-u:-M8HMGZPDYSCZmZiYOgg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9318073c248d3b364ed5244bbd4662e262c29e6/pkg/sql/conn_executor.go#L1535)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Takes some thing thinking to realize that it's OK to put this logic before the update to `txnRewindPos` below. I'd move it after that point.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 1536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M7IjdPX4zv3B3X-tIby:-M8HMH9f6u_hfDsUvsqF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9318073c248d3b364ed5244bbd4662e262c29e6/pkg/sql/conn_executor.go#L1536)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I was just saying to put the `cl.Close()` call in a new `rewindCap.Close()` method.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_test.go, line 815 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H2Kua3PFTNG-CwXla:-M8HMTqMCmlpzV97Z4DO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L815)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

"releases statements" is vague. Say something about that statements buffer.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_test.go, line 820 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H3WXnEJ5LYtsb9kGB:-M8HMZvO4JKLIVgN2jpt:b-8p7co4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L820)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

explain the `4` pls. Better yet, instead of hardcoding something about the way in which some driver uses pgwire, can you iterate through `buf` inside your filter and assert that you never have more than one `SELECT COUNT(*) FROM test` in there?
</blockquote></details>

Done. I chose to record the statement buffer length on the first SELECT COUNT(*) and assert that it doesn't grow.
___
*[pkg/sql/conn_executor_test.go, line 826 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H4ceDAXmCYv8yXS54:-M8KC1zrDHypw0Ze29A4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L826)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this filter takes everything but the command that was just executed. Pls add it, even if you don't use it yet - particularly since with this patch `buf` will no longer contain the respective command. And, in fact, use the command in this test to short-circuit the filter when the command that was just executed is not your `select count`. Otherwise, you're testing all the random executors used by the server - e.g. the internal ones. If you do that, all the atomic stuff can go away since the filter will be synchronized with the one connection triggering it.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_test.go, line 844 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H2gDpCWXytJNb2G-r:-M8K9IcPEMoYuaaE-FI9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L844)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/BeginTx/Begin and the I won't ask you for a comment on the `nil`
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_test.go, line 848 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H33TD8D1mmQOhNdWu:-M8K9IGgA_-y-aEaLQI4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L848)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do you need the `Query()` instead of `Exec()`?
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_test.go, line 850 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8H2lAB3UUryKQsg-sz:-M8HMggQ8XWFWlLG7CDa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c22ab31069372d841015fda7ed631ad2329f1b1/pkg/sql/conn_executor_test.go#L850)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do you need this `for` loop? I think you can remove it.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto commented:
The CI failure is weird. It seems like `stmtBufMaxLen` is updated before the first `SELECT count(*)` query. It's possible that there's another, internal `SELECT count` query that happens before the first one and causes us to update the length before we actually want to. I couldn't reproduce this under stress so I'm just adding a more restrictive filter (the whole statement).

andreimatei reviewed (approved):
LGTM

FWIW, I wouldn't backport this unless some user really wants it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-:-M8LUDI9Da7j4yLvH63p:bg3xufb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)

___
*[pkg/sql/conn_executor_test.go, line 844 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8LTnkR5AXWbbQqfgkS:-M8LTnkR5AXWbbQqfgkT:bnknphu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad0ec583f678c8e6fbbe1887536a33f7d341dcc4/pkg/sql/conn_executor_test.go#L844)):*
> ```Go
> 					// Verify that the buffer doesn't grow.
> 					if l > stmtBufMaxLen {
> 						stmtBufLenOnFailure = l
> ```

I would allow the filter to return an error. Then you don't need `stmtBufLenOnFailure`.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48859)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)

___
*[pkg/sql/conn_executor_test.go, line 844 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48859#-M8LTnkR5AXWbbQqfgkS:-M8LUY8c4aAleiX4qhR4:bfxl9pt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad0ec583f678c8e6fbbe1887536a33f7d341dcc4/pkg/sql/conn_executor_test.go#L844)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I would allow the filter to return an error. Then you don't need `stmtBufLenOnFailure`.
</blockquote></details>

This is assuming that an error returned by `AfterExecCmd` would actually make it to the user, which I guess I'm not sure it would. The connExecutor might interpret it as a communication error and simply close the connection. 


<!-- Sent from Reviewable.io -->

Commit:
pgwire: trim statements past the conn executor's rewind pos eagerly

Previously, the conn executor held on to all executed statements in case they
needed to be retried. These would only be discarded once the transaction
committed. This could lead to a memory blowup in some cases.

With the knowledge that statements whose results have been returned cannot be
retried, this commit eagerly discards those statements every time a command is
executed rather than waiting for the transaction to commit.

Release note (bug fix): executing a large number of statements in a transaction
without committing could previously crash a cockroach server, this is now
fixed.

asubiotto commented:
Added an error to `AfterExecCmd` but that just causes the conn to be closed, as you mentioned. I went back to having a constant that should not be exceeded, but added an accompanying comment. This is to avoid the CI failure which seems to happen because there sometimes isn't a `Sync` command in the statement buffer.

asubiotto commented:
TFTR

bors r=andreimatei

ajwerner commented:
A backport of this seems good. Do you know whether on retries (not automatic retries) we were releasing this memory?

asubiotto commented:
I'm not sure, @andreimatei?

Backporting this

andreimatei commented:
I looked at the time a bit with Andrew and I think it looked like we weren't releasing the memory, but I wasn't sure.
